core/src/main/scala/spire/optional/Perm.scala

```diff
@@ -0,0 +1,62 @@
+package spire
+package optional
+
+import scala.collection.compat._
+import spire.scalacompat.SeqLike
+
+import cats.kernel.Eq
+import spire.algebra.{Action, Group}
+import spire.algebra.partial.PartialAction
+import spire.syntax.cfor._
+import spire.syntax.group._
+import spire.util._
+
+/**
+ * Represents a permutation encoded as a map from preimages to images, including
+ * only pairs that are moved by the permutation (so the identity is Map.empty).
+ */
+case class Perm(mapping: Map[Int, Int]) extends AnyVal {
+  def apply(k: Int): Int = mapping.getOrElse(k, k)
+  def image: Set[Int] = mapping.keySet
+  def invert(k: Int): Int = mapping.find(_._2 == k).fold(k)(_._1)
+}
+
+object Perm {
+  implicit val PermEq: Eq[Perm] = Eq.fromUniversalEquals
+  implicit val PermIntAction: Action[Int, Perm] = new PermIntAction
+  implicit val PermGroup: Group[Perm] = new PermGroup
+  implicit def PermSeqPartialAction[A, CC[A] <: SeqLike[A, CC[A]]](implicit cbf: Factory[A, CC[A]]): PartialAction[CC[A], Perm] = new PermSeqPartialAction[A, CC[A]]
+}
+
+final class PermIntAction extends Action[Int, Perm] {
+  def actr(k: Int, perm: Perm): Int = perm(k)
+  def actl(perm: Perm, k: Int): Int = perm.invert(k)
+}
+
+final class PermGroup extends Group[Perm] {
+  def empty: Perm = Perm(Map.empty[Int, Int])
+  def combine(x: Perm, y: Perm): Perm = Perm {
+    val preimages = x.image | y.image
```

Comment:
Subtle change. Formerly concatenated, which adds every pair redundantly in the worst case scenario. (A common case, if you're combining two permutations from the same symmetric group).